lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -834,20 +834,19 @@ _maxScriptExecutionCost
     :: ProtocolParameters
     -> [Redeemer]
     -> Coin
-_maxScriptExecutionCost pp (length -> numberOfScripts)  =
-    case view #executionUnitPrices pp of
-        Just prices -> totalCost $ executionCost prices maxExecutionUnits
+_maxScriptExecutionCost pp redeemers
+    | not (null redeemers) = case view #executionUnitPrices pp of
+        Just prices -> executionCost prices maxExecutionUnits
         Nothing     -> Coin 0
+    | otherwise            = Coin 0
   where
     maxExecutionUnits :: ExecutionUnits
     maxExecutionUnits = view (#txParameters . #getMaxExecutionUnits) pp
 
-    totalCost :: Rational -> Coin
-    totalCost = Coin.unsafeNaturalToCoin . ceiling . (* (fromIntegral numberOfScripts) )
-
-    executionCost :: ExecutionUnitPrices -> ExecutionUnits -> Rational
+    executionCost :: ExecutionUnitPrices -> ExecutionUnits -> Coin
     executionCost (ExecutionUnitPrices perStep perMem) (W.ExecutionUnits steps mem) =
-        perStep * (toRational steps) + perMem * (toRational mem)
+        Coin.unsafeNaturalToCoin . ceiling $
+            perStep * (toRational steps) + perMem * (toRational mem)
```

Comment:
I found this code a little hard to understand, since the field names used in this function are quite different from the field names of the original record types. I think it would be good to use consistent terminology everywhere, to help people trying to understand the code in future.

Looking at the types themselves, I think I prefer **_your_** names!

Can we change the field names of `ExecutionUnitPrices` to something like:
```patch
data ExecutionUnitPrices = ExecutionUnitPrices
-     { priceExecutionSteps
+     { pricePerStep
          :: Rational
-     , priceExecutionMemory
+     , pricePerMemoryUnit
          :: Rational
      } deriving (Eq, Generic, Show)
```
I think calling it `perMemoryUnit` is also clearer from a plain English perspective (since "per memory" is a bit ambiguous).

Secondly, could we rewrite this function to use labels? Attempting to pattern match these records on a single line makes that line rather long. (This approach doesn't really scale if you have more parameters.)

So perhaps:
```hs
executionCost :: ExecutionUnitPrices -> ExecutionUnits -> Coin
executionCost ps us = Coin.unsafeNaturalToCoin . ceiling
    $ (ps ^. #pricePerStep)       * toRational (us ^. #steps)
    + (ps ^. #pricePerMemoryUnit) * toRational (us ^. #memoryUnits)
```

Reply:
Update: I made a suggested commit here, which you could use if you prefer:

20273ebb3b9c8601000b2e07e501f518fcda874f

Reply:
Thanks, I think the renaming definitely makes sense.

> Secondly, could we rewrite this function to use labels?

You do lose the ability to spot unused fields… but sure, its neater in other ways